File: rust/src/bin/p0042.rs
// Project Euler: Problem 42

euler::run_solver!(42);

static FILE_DATA: &str = include_str!("../../assets/0042_words.txt");

fn solve() -> String {
    compute(FILE_DATA).to_string()
}

fn compute(data: &str) -> usize {
    use euler::math;

    parse_data(data)
        .into_iter()
        .filter(|s| math::is_triangular(worth(s)))
        .count()
}

fn parse_data(s: &str) -> Vec<String> {
    use std::string::ToString;

    s.chars()
        .filter(|&c| c != '"')
        .collect::<String>()
        .split(',')
        .map(ToString::to_string)
        .collect()
}

fn worth(s: &str) -> i64 {
    s.chars().map(|c| c as i64 - 'A' as i64 + 1).sum()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0042() {
        assert_eq!(compute(super::FILE_DATA), 162);
    }
}

File: ocaml/src/p0042.ml
(* Project Euler: Problem 42 *)

open Core

let parse_data data =
  let line = List.hd_exn data in
  List.(
    String.sub line ~pos:1 ~len:(String.length line - 2)
    (* If the Core library supports `split_all` or an equivalent feature, I'll use it. *)
    |> Stdlib.String.split_all ~sep:{|","|}
    |> map ~f:String.to_list
    |> map ~f:(sum (module Int) ~f:(fun ch -> Char.to_int ch - 0x40)))
;;

let compute str_lst = parse_data str_lst |> List.count ~f:Euler.Math.is_triangular
let solve fname = compute (Euler.Task.read_file fname) |> Int.to_string

(* Test *)

let%test_unit "p042_words.txt" =
  [%test_eq: int] (compute (Euler.Task.read_file "./assets/p042_words.txt")) 162
;;
